panic when parsing `func{{a}}`

**Describe the bug**

cel-go will panic if the input is ``func{{a}}``

**To Reproduce**
Check which components this affects:

- [x]  parser
- [ ]  checker
- [ ]  interpreter

Sample expression and input that reproduces the issue:
```cel
func{{a}}
```

add 

```
{
		I: `func{{a}}`,
		E: ``,
},
```

to `testCases` in `parser/parser_test.go` and run `TestParse`

```
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 35 [running]:
testing.tRunner.func1(0xc00014a200)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x15d05e0, 0xc00003a240)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/google/cel-go/parser.(*parser).VisitIFieldInitializerList(0xc00000e060, 0x16fcc20, 0xc00019e360, 0x4, 0xc000033601)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:315 +0x41c
github.com/google/cel-go/parser.(*parser).VisitCreateMessage(0xc00000e060, 0xc00019b890, 0xc00019b890, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:301 +0x12d
github.com/google/cel-go/parser.(*parser).VisitMemberExpr(0xc00000e060, 0xc000011900, 0x100bde5, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:231 +0x1ee
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16efa80, 0xc000011900, 0x16efa80, 0xc000011900)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:106 +0x686
github.com/google/cel-go/parser.(*parser).VisitCalc(0xc00000e060, 0xc00019b050, 0x100bde5, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:202 +0x381
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16ef580, 0xc00019b050, 0x16ef580, 0xc00019b050)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:102 +0x4d2
github.com/google/cel-go/parser.(*parser).VisitRelation(0xc00000e060, 0xc00019b020, 0x16ec940, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:184 +0x381
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16efc00, 0xc00019b020, 0x16efc00, 0xc00019b020)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:100 +0x36e
github.com/google/cel-go/parser.(*parser).VisitConditionalAnd(0xc00000e060, 0xc000168a80, 0x8, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:168 +0x7e
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16ef600, 0xc000168a80, 0x16ef600, 0xc000168a80)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:96 +0x6e2
github.com/google/cel-go/parser.(*parser).VisitConditionalOr(0xc00000e060, 0xc000168a00, 0x4, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:153 +0x7e
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16ef680, 0xc000168a00, 0x16ef680, 0xc000168a00)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:98 +0x52e
github.com/google/cel-go/parser.(*parser).VisitExpr(0xc00000e060, 0xc0000b1a40, 0x100bde5, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:141 +0x6a
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16ef880, 0xc0000b1a40, 0x16ef880, 0xc0000b1a40)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:94 +0x442
github.com/google/cel-go/parser.(*parser).VisitStart(0xc00000e060, 0xc00019adb0, 0x8, 0x100ac9b)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:136 +0x5d
github.com/google/cel-go/parser.(*parser).Visit(0xc00000e060, 0x16efd00, 0xc00019adb0, 0x16efd00, 0xc00019adb0)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:92 +0x5a6
github.com/google/cel-go/parser.(*parser).parse(0xc00000e060, 0xc00003e190, 0x9, 0xd)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:85 +0x333
github.com/google/cel-go/parser.ParseWithMacros(0x16edb40, 0xc000158000, 0x1b236e0, 0x7, 0x7, 0xc000158000, 0x4aec1712620ff)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:59 +0x237
github.com/google/cel-go/parser.Parse(...)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser.go:45
github.com/google/cel-go/parser.TestParse.func1(0xc00014a200)
	/Users/virusdefender/Desktop/workspace/cel-go/parser/parser_test.go:125 +0xb8
testing.tRunner(0xc00014a200, 0xc0000e4cd0)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350
```

**Expected behavior**

the parse result is syntax error

--- a/parser/parser.go
+++ b/parser/parser.go
@@ -323,11 +323,7 @@ func (p *parser) VisitIFieldInitializerList(ctx gen.IFieldInitializerListContext
 	cols := ctx.GetCols()
 	vals := ctx.GetValues()
 	for i, f := range ctx.GetFields() {
-		if i >= len(cols) {
-			// This is the result of a syntax error detected elsewhere.
-			return []*exprpb.Expr_CreateStruct_Entry{}
-		}
-		if i >= len(vals) {
+		if i >= len(cols) || i >= len(vals) {
 			// This is the result of a syntax error detected elsewhere.
 			return []*exprpb.Expr_CreateStruct_Entry{}
 		}
@@ -420,11 +416,7 @@ func (p *parser) VisitMapInitializerList(ctx *gen.MapInitializerListContext) int
 	vals := ctx.GetValues()
 	for i, col := range ctx.GetCols() {
 		colID := p.helper.id(col)
-		if i >= len(keys) {
-			// This is the result of a syntax error detected elsewhere.
-			return []*exprpb.Expr_CreateStruct_Entry{}
-		}
-		if i >= len(vals) {
+		if i >= len(keys) || i >= len(vals) {
 			// This is the result of a syntax error detected elsewhere.
 			return []*exprpb.Expr_CreateStruct_Entry{}
 		}

--- a/parser/parser.go
+++ b/parser/parser.go
@@ -155,8 +155,12 @@ func (p *parser) VisitConditionalOr(ctx *gen.ConditionalOrContext) interface{} {
 		return result
 	}
 	b := newBalancer(p.helper, operators.LogicalOr, result)
+	rest := ctx.GetE1()
 	for i, op := range ctx.GetOps() {
-		next := p.Visit(ctx.GetE1()[i]).(*exprpb.Expr)
+		if i >= len(rest) {
+			return p.reportError(ctx, "unexpected character, wanted '||'")
+		}
+		next := p.Visit(rest[i]).(*exprpb.Expr)
 		opID := p.helper.id(op)
 		b.addTerm(opID, next)
 	}
@@ -170,8 +174,12 @@ func (p *parser) VisitConditionalAnd(ctx *gen.ConditionalAndContext) interface{}
 		return result
 	}
 	b := newBalancer(p.helper, operators.LogicalAnd, result)
+	rest := ctx.GetE1()
 	for i, op := range ctx.GetOps() {
-		next := p.Visit(ctx.GetE1()[i]).(*exprpb.Expr)
+		if i >= len(rest) {
+			return p.reportError(ctx, "unexpected character, wanted '&&'")
+		}
+		next := p.Visit(rest[i]).(*exprpb.Expr)
 		opID := p.helper.id(op)
 		b.addTerm(opID, next)
 	}
@@ -307,13 +315,24 @@ func (p *parser) VisitCreateMessage(ctx *gen.CreateMessageContext) interface{} {
 // Visit a parse tree of field initializers.
 func (p *parser) VisitIFieldInitializerList(ctx gen.IFieldInitializerListContext) interface{} {
 	if ctx == nil || ctx.GetFields() == nil {
+		// This is the result of a syntax error handled elswhere, return empty.
 		return []*exprpb.Expr_CreateStruct_Entry{}
 	}
 
 	result := make([]*exprpb.Expr_CreateStruct_Entry, len(ctx.GetFields()))
+	cols := ctx.GetCols()
+	vals := ctx.GetValues()
 	for i, f := range ctx.GetFields() {
-		initID := p.helper.id(ctx.GetCols()[i])
-		value := p.Visit(ctx.GetValues()[i]).(*exprpb.Expr)
+		if i >= len(cols) {
+			// This is the result of a syntax error detected elsewhere.
+			return []*exprpb.Expr_CreateStruct_Entry{}
+		}
+		if i >= len(vals) {
+			// This is the result of a syntax error detected elsewhere.
+			return []*exprpb.Expr_CreateStruct_Entry{}
+		}
+		initID := p.helper.id(cols[i])
+		value := p.Visit(vals[i]).(*exprpb.Expr)
 		field := p.helper.newObjectField(initID, f.GetText(), value)
 		result[i] = field
 	}
@@ -389,31 +408,28 @@ func (p *parser) VisitConstantLiteral(ctx *gen.ConstantLiteralContext) interface
 	return p.reportError(ctx, "invalid literal")
 }
 
-// Visit a parse tree produced by CELParser#exprList.
-func (p *parser) VisitExprList(ctx *gen.ExprListContext) interface{} {
-	if ctx == nil || ctx.GetE() == nil {
-		return []*exprpb.Expr{}
-	}
-
-	result := make([]*exprpb.Expr, len(ctx.GetE()))
-	for i, e := range ctx.GetE() {
-		exp := p.Visit(e).(*exprpb.Expr)
-		result[i] = exp
-	}
-	return result
-}
-
 // Visit a parse tree produced by CELParser#mapInitializerList.
 func (p *parser) VisitMapInitializerList(ctx *gen.MapInitializerListContext) interface{} {
 	if ctx == nil || ctx.GetKeys() == nil {
+		// This is the result of a syntax error handled elswhere, return empty.
 		return []*exprpb.Expr_CreateStruct_Entry{}
 	}
 
 	result := make([]*exprpb.Expr_CreateStruct_Entry, len(ctx.GetCols()))
+	keys := ctx.GetKeys()
+	vals := ctx.GetValues()
 	for i, col := range ctx.GetCols() {
 		colID := p.helper.id(col)
-		key := p.Visit(ctx.GetKeys()[i]).(*exprpb.Expr)
-		value := p.Visit(ctx.GetValues()[i]).(*exprpb.Expr)
+		if i >= len(keys) {
+			// This is the result of a syntax error detected elsewhere.
+			return []*exprpb.Expr_CreateStruct_Entry{}
+		}
+		if i >= len(vals) {
+			// This is the result of a syntax error detected elsewhere.
+			return []*exprpb.Expr_CreateStruct_Entry{}
+		}
+		key := p.Visit(keys[i]).(*exprpb.Expr)
+		value := p.Visit(vals[i]).(*exprpb.Expr)
 		entry := p.helper.newMapEntry(colID, key, value)
 		result[i] = entry
 	}

--- a/parser/parser_test.go
+++ b/parser/parser_test.go
@@ -1019,6 +1019,48 @@ ERROR: <input>:1:14: argument is not an identifier
  | [1, 2, 3].map(var, var * var)
  | .............^`,
 	},
+	{
+		I: "func{{a}}",
+		E: `ERROR: <input>:1:6: Syntax error: extraneous input '{' expecting {'}', IDENTIFIER}
+		| func{{a}}
+		| .....^
+	   ERROR: <input>:1:8: Syntax error: mismatched input '}' expecting ':'
+		| func{{a}}
+		| .......^
+	   ERROR: <input>:1:9: Syntax error: extraneous input '}' expecting <EOF>
+		| func{{a}}
+		| ........^`,
+	},
+	{
+		I: "msg{:a}",
+		E: `ERROR: <input>:1:5: Syntax error: extraneous input ':' expecting {'}', IDENTIFIER}
+		| msg{:a}
+		| ....^
+	   ERROR: <input>:1:7: Syntax error: mismatched input '}' expecting ':'
+		| msg{:a}
+		| ......^`,
+	},
+	{
+		I: "{a}",
+		E: `ERROR: <input>:1:3: Syntax error: mismatched input '}' expecting ':'
+		| {a}
+		| ..^`,
+	},
+	{
+		I: "{:a}",
+		E: `ERROR: <input>:1:2: Syntax error: extraneous input ':' expecting {'[', '{', '}', '(', '.', '-', '!', 'true', 'false', 'null', NUM_FLOAT, NUM_INT, NUM_UINT, STRING, BYTES, IDENTIFIER}
+		| {:a}
+		| .^
+	   ERROR: <input>:1:4: Syntax error: mismatched input '}' expecting ':'
+		| {:a}
+		| ...^`,
+	},
+	{
+		I: "ind[a{b}]",
+		E: `ERROR: <input>:1:8: Syntax error: mismatched input '}' expecting ':'
+		| ind[a{b}]
+		| .......^`,
+	},
 }
 
 type testInfo struct {

